core/storage/suggestion/gae_models.py

```diff
@@ -2408,11 +2418,41 @@ def fetch_page(
                 if not result_models:
                     break
                 for result_model in result_models:
+                    if len(sorted_results) == page_size:
+                        break
+                    next_offset += 1
+                    if (((result_model.first_contribution_date) <= (
+                        first_date)) and (
+                        result_model.last_contribution_date >= last_date)):
+                        sorted_results.append(result_model)
+        elif max_days_since_first_activity is not None:
+            next_offset = offset
+            while len(sorted_results) < page_size:
+                result_models = (
+                    sort_query.fetch(
+                        NUM_MODELS_PER_FETCH, offset=next_offset))
+                if not result_models:
+                    break
+                for result_model in result_models:
+                    if len(sorted_results) == page_size:
+                        break
+                    next_offset += 1
+                    if result_model.first_contribution_date <= first_date:
+                        sorted_results.append(result_model)
+        elif max_days_since_last_activity is not None:
+            next_offset = offset
+            while len(sorted_results) < page_size:
+                result_models = (
+                    sort_query.fetch(
+                        NUM_MODELS_PER_FETCH, offset=next_offset))
+                if not result_models:
+                    break
+                for result_model in result_models:
+                    if len(sorted_results) == page_size:
+                        break
                     next_offset += 1
                     if result_model.last_contribution_date >= last_date:
                         sorted_results.append(result_model)
```

Comment:
There seems to be a lot of code duplication going on here. It helps to start by noting which lines of code are the same between conditional branches and which lines if code are different between conditional branches. Ditto below.

Reply:
Actually nvm. We'll push refactoring this for later.